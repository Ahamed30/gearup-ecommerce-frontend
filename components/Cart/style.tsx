import { openSans, rubik } from "@/styles";

const fontRubik = rubik?.className;
const fontOpenSans = openSans?.className;

export const saleTitleClassName = `${fontRubik} text-3xl font-semibold color-[#232321] mb-[8px]`;
export const saleExcerptClassName = `${fontOpenSans} text-sm font-semibold color-[#232321] mb-[8px]`;
export const linkClassName = `${fontOpenSans} text-base font-semibold color-[#232321] mb-[8px]`;
