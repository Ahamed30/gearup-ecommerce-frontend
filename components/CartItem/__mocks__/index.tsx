export const cartItemData = {
  id: "clk8dhn0s0udx0bpbmqrkru3f",
  quantity: 1,
  heroImage: {
    id: "clk8didmw0ug60bpbkp00pcrv",
    url: "https://media.graphassets.com/tEoKFf1JQ1a2QsVIOTdH",
  },
  productName: "ADIDAS 4DFWD X PARLEY RUNNING SHOES",
  price: 125,
  salePrice: 100,
  size: "10",
  color: "Yellow",
};
