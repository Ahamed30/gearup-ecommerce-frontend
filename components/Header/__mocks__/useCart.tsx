// __mocks__/useCart.ts
import { createContext, useContext } from "react";

// Create a mock cart context with a default value of null (no cart data).
const CartContext = createContext(null);

// Create a custom hook to access the mock cart context.
export const useCart = () => {
  const cart = useContext(CartContext);
  if (cart === null) {
    throw new Error("useCart must be used within a CartProvider");
  }
  return cart;
};
