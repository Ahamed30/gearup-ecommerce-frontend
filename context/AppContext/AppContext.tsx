import { type ReactNode, useContext, useState, createContext } from "react";

export interface AppContextType {
  isLoading: boolean;
  // eslint-disable-next-line no-unused-vars
  setIsLoading: (isLoading: boolean) => void;
}

//@ts-expect-error: ignore initial context creation
const AppContext = createContext<AppContextType>(null);

export const useApp = () => useContext(AppContext);

export interface AppContextProviderProps {
  children: ReactNode;
}
export const AppContextProvider = ({ children }: AppContextProviderProps) => {
  const [isLoading, setIsLoading] = useState(false);

  return (
    <AppContext.Provider value={{ isLoading, setIsLoading }}>
      {children}
    </AppContext.Provider>
  );
};
