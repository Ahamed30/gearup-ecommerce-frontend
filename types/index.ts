export * from "./header";
export * from "./link";
export * from "./footer";
export * from "./linkColumn";
export * from "./subscribeForm";
export * from "./asset";
export * from "./homePage";
export * from "./productCard";
