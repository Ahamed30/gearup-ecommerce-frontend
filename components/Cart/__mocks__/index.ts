import { CartType } from "@/types";

export const cartWithItems: CartType = {
  items: [
    {
      id: "clk8dhn0s0udx0bpbmqrkru3f",
      quantity: 1,
      heroImage: {
        id: "clk8didmw0ug60bpbkp00pcrv",
        url: "https://media.graphassets.com/tEoKFf1JQ1a2QsVIOTdH",
      },
      productName: "ADIDAS 4DFWD X PARLEY RUNNING SHOES",
      price: 125,
      salePrice: 100,
      size: "10",
      color: "Yellow",
    },
    {
      id: "clk29a3wv2o250bpqy9fxnqxq",
      quantity: 1,
      heroImage: {
        id: "clk29209r2oap0bnxkd7aqmi3",
        url: "https://media.graphassets.com/QB2jGvWSqGxI4J0IZMpj",
      },
      productName: "Nike Air Max Plus III",
      price: 170,
      salePrice: 120,
      size: "8",
      color: "Blue",
    },
  ],
  itemCount: 2,
  cartSubTotal: 270,
};
