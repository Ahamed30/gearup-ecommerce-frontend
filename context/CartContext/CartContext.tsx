import { createContext, useContext, useState } from "react";
import type { ReactNode, Key } from "react";
import { Cart, CartItemType } from "@/types";
import { getCookie, removeCookie, setCookie } from "@/utils/cookie";

interface CartContextType {
  addToCart: (item: CartItemType) => void; // when updating via backend these can be update to Promise<void> from void
  cart: Cart | null;
  // clearCartAlertState: () => void;
  removeFromCart: (itemId: Key, size: string) => void;
  updateCartItem: (itemId: string, updatedItem: CartItemType) => void;
  updateCart: (cart: Cart) => void;
  emptyCart: () => void;
}

//@ts-expect-error: ignore initial context creation
const CartContext = createContext<CartContextType>(null);

export const useCart = () => useContext(CartContext);

export interface CartContextProviderProps {
  cart?: Cart | null;
  children: ReactNode;
}

export const CartContextProvider = ({
  cart: initialCart = null,
  children,
}: CartContextProviderProps) => {
  const cartFromCookie: Cart | null = getCookie("cart");
  // console.log("...", cartFromCookie);
  const [cart, setCart] = useState(initialCart ?? cartFromCookie);

  const addToCart = (item: CartItemType) => {
    let isItemExists = false;
    if (cart?.items) {
      cart.items.forEach((element) => {
        if (element.id === item.id && element.size === item.size) {
          cart.cartSubTotal =
            (cart.cartSubTotal ?? 0) + (item?.salePrice ?? item.price);
          element.quantity += 1;
          isItemExists = true;
          setCookie("cart", cart);
          return;
        }
      });
    }
    if (isItemExists) return;
    setCart((prevCart) => {
      const updatedCart: Cart = { ...prevCart };

      updatedCart.items = [...(prevCart?.items || []), item];
      updatedCart.itemCount = (updatedCart.itemCount ?? 0) + 1;
      updatedCart.cartSubTotal =
        (updatedCart.cartSubTotal ?? 0) + (item?.salePrice ?? item.price);
      setCookie("cart", updatedCart);
      return updatedCart;
    });
  };

  const removeFromCart = (itemId: Key, size: string) => {
    setCart((prevCart) => {
      const updatedCart: Cart = { ...prevCart };
      updatedCart.items = updatedCart.items?.filter((item) => {
        if (item.id === itemId && item.size === size) {
          updatedCart.cartSubTotal =
            (updatedCart.cartSubTotal ?? 0) -
            (item?.salePrice ?? item.price) * item.quantity;
          return false;
        }
        return true;
      });

      updatedCart.itemCount = (updatedCart.itemCount ?? 1) - 1;
      setCookie("cart", updatedCart);
      return updatedCart;
    });
    console.log("-->", cart);
  };

  const updateCartItem = (itemId: Key, updatedItem: CartItemType) => {
    setCart((prevCart) => {
      const updatedCart: Cart = { ...prevCart };

      updatedCart.items = updatedCart.items?.map((item) =>
        item.id === itemId ? updatedItem : item
      );
      setCookie("cart", updatedCart);
      return updatedCart;
    });
  };

  const updateCart = (cart: Cart) => {
    setCart(cart);
    setCookie("cart", cart);
  };

  const emptyCart = () => {
    removeCookie("cart");
    setCart(null);
  };

  return (
    <CartContext.Provider
      value={{
        cart,
        addToCart,
        removeFromCart,
        updateCartItem,
        updateCart,
        emptyCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};
