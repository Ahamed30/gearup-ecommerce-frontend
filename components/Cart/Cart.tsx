"use client";

import Link from "next/link";
import { CartItem } from "../CartItem";
import { Typography } from "../Typography";
import { OrderSummary } from "../OrderSummary";
import { useCart } from "@/context/CartContext";
import { useState } from "react";
import { Modal } from "../Modal";

export const Cart = () => {
  const { cart, emptyCart } = useCart();
  const [showModal, setShowModal] = useState<boolean>(false);

  return (
    <div className="mb-[32px]">
      <div className="mb-[24px]">
        <Typography
          variant="headline"
          className="text-3xl font-semibold mb-[8px]"
        >
          Saving to celebrate
        </Typography>
        <Typography className="text-sm font-semibold mb-[8px]">
          Enjoy up to 60% off thousands of styles during the End of Year sale -
          while suppiles last. No code needed.
        </Typography>
        <Typography className="text-base font-semibold mb-[8px]">
          <Link href="/" className="underline">
            Join us
          </Link>{" "}
          or{" "}
          <Link href="/" className="underline">
            Sign-in
          </Link>
        </Typography>
      </div>
      <div className="flex flex-col lg:flex-row gap-8">
        <div className="w-full lg:w-3/5">
          <div className="border rounded-[16px] bg-[#FAFAFA] p-4 lg:p-6">
            <div className="mb-[32px]">
              {!cart?.itemCount ? (
                <div className="flex justify-center items-center">
                  <Typography
                    variant="headline"
                    className="text-xl lg:text-3xl font-semibold mb-[8px]"
                  >
                    Your Bag is Empty!!!!!
                  </Typography>
                </div>
              ) : (
                <>
                  <div className="flex justify-between">
                    <Typography
                      variant="headline"
                      className="text-xl lg:text-3xl font-semibold mb-[8px]"
                    >
                      Your Bag
                    </Typography>
                    <Typography
                      className="underline underline-offset-3 cursor-pointer"
                      onClick={() => setShowModal(true)}
                    >
                      Empty Cart
                    </Typography>
                    {showModal && (
                      <Modal
                        title="Empty Cart"
                        subText="Are you sure you want to empty the cart?"
                        setShowModal={setShowModal}
                        handleClick={emptyCart}
                      />
                    )}
                  </div>
                  <Typography className="text-sm lg:text-base font-normal">
                    Items in your bag not reserved- check out now to make them
                    yours.
                  </Typography>
                </>
              )}
            </div>
            <div>
              {cart?.items?.map((item) => {
                return <CartItem key={item?.id} product={item} />;
              })}
            </div>
          </div>
        </div>
        <OrderSummary step="cart" />
      </div>
    </div>
  );
};
