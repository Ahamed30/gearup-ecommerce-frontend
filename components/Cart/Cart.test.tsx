import { fireEvent, render, screen } from "@testing-library/react";
import { Cart } from "./Cart";
import { CartContextProvider } from "@/context/CartContext";
import { cartWithItems } from "./__mocks__";
import { CartType } from "@/types";
import { AppContextProvider } from "@/context/AppContext";

const mockRouterPush = jest.fn();
jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: mockRouterPush,
  }),
}));

const renderComponent = (cart: CartType = cartWithItems) => {
  return render(
    <AppContextProvider>
      <CartContextProvider cart={cart}>
        <Cart />
      </CartContextProvider>
    </AppContextProvider>
  );
};

describe("Cart", () => {
  test("should match the snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("renders cart items when there are items in the cart", () => {
    renderComponent();
    const cartItems = screen.getAllByRole("cartItem");
    expect(cartItems.length).toBe(2);
  });

  test("should render a message indicating an empty cart", () => {
    const emptyCart: CartType = {
      items: [],
      itemCount: 0,
      cartSubTotal: 0,
    };
    renderComponent(emptyCart);
    const element = screen.getByText("Your Bag is Empty!!!!!");
    expect(element).toBeInTheDocument();
  });

  test("should open the empty cart modal when the Empty Cart button is clicked and clears cart", () => {
    renderComponent();
    const emptyCartButton = screen.getByText("Empty Cart");
    fireEvent.click(emptyCartButton);
    const dialogBox = screen.getByRole("dialog");
    expect(dialogBox).toBeInTheDocument();
    const yesButton = screen.getByText("Yes, I'm sure");
    fireEvent.click(yesButton);
    const element = screen.getByText("Your Bag is Empty!!!!!");
    expect(element).toBeInTheDocument();
  });
});
