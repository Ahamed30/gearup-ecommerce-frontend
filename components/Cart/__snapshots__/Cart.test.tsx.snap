// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should match the snapshot 1`] = `
<div>
  <div
    class="mb-[32px]"
  >
    <div
      class="mb-[24px]"
    >
      <p
        class="className text-3xl font-semibold mb-[8px]"
        style="color: rgb(35, 35, 33);"
      >
        Saving to celebrate
      </p>
      <p
        class="className text-sm font-semibold mb-[8px]"
        style="color: rgb(35, 35, 33);"
      >
        Enjoy up to 60% off thousands of styles during the End of Year sale - while suppiles last. No code needed.
      </p>
      <p
        class="className text-base font-semibold mb-[8px]"
        style="color: rgb(35, 35, 33);"
      >
        <a
          class="underline"
          href="/"
        >
          Join us
        </a>
         
        or
         
        <a
          class="underline"
          href="/"
        >
          Sign-in
        </a>
      </p>
    </div>
    <div
      class="flex flex-col lg:flex-row gap-8"
    >
      <div
        class="w-full lg:w-3/5"
      >
        <div
          class="border rounded-[16px] bg-[#FAFAFA] p-4 lg:p-6"
        >
          <div
            class="mb-[32px]"
          >
            <div
              class="flex justify-center items-center"
            >
              <p
                class="className text-xl lg:text-3xl font-semibold mb-[8px]"
                style="color: rgb(35, 35, 33);"
              >
                Your Bag is Empty!!!!!
              </p>
            </div>
          </div>
          <div />
        </div>
      </div>
      <div
        class="border rounded-[16px] bg-[#FAFAFA] lg:flex-grow p-6 lg:border-none lg:rounded-none lg:bg-transparent"
      >
        <div>
          <p
            class="className text-xl lg:text-2xl font-semibold mb-[16px] lg:mb-[24px]"
            style="color: rgb(35, 35, 33);"
          >
            Order Summary
          </p>
          <div
            class="flex justify-between mb-[16px]"
          >
            <p
              class="className text-base lg:text-lg font-semibold"
              style="color: rgb(35, 35, 33);"
            >
              SubTotal
            </p>
            <p
              class="className text-base lg:text-lg font-semibold"
              style="color: rgb(35, 35, 33);"
            >
              0
            </p>
          </div>
          <div
            class="flex justify-between mb-[16px]"
          >
            <p
              class="className text-base lg:text-lg font-semibold"
              style="color: rgb(35, 35, 33);"
            >
              Estimated Tax
            </p>
            <p
              class="className text-base lg:text-lg font-semibold"
              style="color: rgb(35, 35, 33);"
            >
              -
            </p>
          </div>
          <div
            class="flex justify-between mb-[16px] lg:mb-[24px]"
          >
            <p
              class="className text-lg lg:text-xl font-semibold"
              style="color: rgb(35, 35, 33);"
            >
              Total
            </p>
            <p
              class="className text-lg lg:text-xl font-semibold"
              style="color: rgb(35, 35, 33);"
            >
              0
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
